File: 061. Rotate List/list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == nullptr)
            return nullptr;
        ListNode *dummy = new ListNode(-1), *p = head, *tail = nullptr;
        // count length
        int len = 0;
        while(p){
            p = p->next;
            len++;
        }
        k = k % len;
        if(k == 0) 
            return head;
        // find tail
        p = head;
        for(int i = 0; i < len - k - 1; ++i){
            p = p->next;
        }
        // split it into two lists
        ListNode* rest = p->next;
        p->next = nullptr;
        // merge
        dummy->next = rest;
        while(rest->next){
            rest = rest->next;
        }
        rest->next = head;

        return dummy->next;
    }
};

File: offer/JZ25 合并两个排序的链表/list.java
/*
public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode Merge(ListNode l1,ListNode l2) {
        ListNode dummy = new ListNode(0), tail = dummy;
        while(l1 != null && l2 != null){
            if(l1.val <= l2.val){
                tail.next = l1;
                l1 = l1.next;
                tail = tail.next;
                tail.next = null;
            }else{
                tail.next = l2;
                l2 = l2.next;
                tail = tail.next;
                tail.next = null;
            }
        }
        if(l1 != null)
            tail.next = l1;
        if(l2 != null)
            tail.next = l2;
        return dummy.next;
    }
}